File: code/nodejs/03-agent-loop/tool_dispatcher.ts
/**
 * Tool execution dispatcher for the ReAct agent loop.
 *
 * dispatchTool() is the bridge between the LLM's tool_call decision and your
 * TypeScript functions. It:
 *   - Looks up the tool by name in a registry
 *   - Parses the JSON arguments from the tool_call
 *   - Calls the function and returns a formatted tool message
 *   - Returns a descriptive error message (not a thrown exception) if anything
 *     fails — the LLM receives the error and can explain it to the user
 *
 * See docs/02-the-agent-loop/01-anatomy-of-an-agent.md — "The Hands (Tools)"
 */

import type {
  ChatCompletionMessageParam,
  ChatCompletionMessageToolCall,
} from "openai/resources/chat/completions";

// Registry maps tool names to functions that accept parsed args.
export type ToolFunction = (args: Record<string, unknown>) => unknown;
export type ToolRegistry = Record<string, ToolFunction>;

/**
 * Execute a tool call and return a formatted tool message.
 *
 * @param toolCall - An OpenAI tool call object with .id, .function.name,
 *                   and .function.arguments (JSON string).
 * @param registry - Mapping of tool names to callables.
 * @returns A tool message ready to push onto the messages array:
 *          `{ role: "tool", content: "<json>", tool_call_id: "<id>" }`
 */
export function dispatchTool(
  toolCall: ChatCompletionMessageToolCall,
  registry: ToolRegistry
): ChatCompletionMessageParam {
  const { id: toolCallId, function: fn } = toolCall;
  const name = fn.name;

  // Parse arguments.
  let args: Record<string, unknown>;
  try {
    args = JSON.parse(fn.arguments) as Record<string, unknown>;
  } catch (err) {
    console.error(`[dispatchTool] Tool ${name}: invalid argument JSON:`, err);
    return _errorMessage(toolCallId, `Invalid arguments JSON: ${String(err)}`);
  }

  // Look up the tool.
  const toolFn = registry[name];
  if (!toolFn) {
    const available = Object.keys(registry).join(", ");
    console.warn(`[dispatchTool] Tool '${name}' not found. Available: ${available}`);
    return _errorMessage(
      toolCallId,
      `Tool '${name}' is not available. Available tools: ${available}`
    );
  }

  // Execute with timing.
  const start = performance.now();
  let result: unknown;
  try {
    result = toolFn(args);
  } catch (err) {
    const elapsed = (performance.now() - start).toFixed(1);
    console.error(`[dispatchTool] Tool '${name}' failed (${elapsed} ms):`, err);
    return _errorMessage(toolCallId, `Tool '${name}' failed: ${String(err)}`);
  }

  const elapsed = (performance.now() - start).toFixed(1);
  const content = JSON.stringify(result);
  const preview = content.length > 200 ? content.slice(0, 200) + "…" : content;
  console.debug(
    `[dispatchTool] ${name}(${JSON.stringify(args)}) → ${preview}  [${elapsed} ms]`
  );

  return {
    role: "tool",
    content,
    tool_call_id: toolCallId,
  };
}

function _errorMessage(toolCallId: string, error: string): ChatCompletionMessageParam {
  return {
    role: "tool",
    content: JSON.stringify({ error }),
    tool_call_id: toolCallId,
  };
}

File: code/python/03-agent-loop/tool_dispatcher.py
"""Tool execution dispatcher.

dispatch_tool() is the bridge between the LLM's tool_call decision and your
Python functions. It:

  - Looks up the tool by name in a registry dict
  - Parses the JSON arguments string from the tool_call
  - Calls the Python function with keyword arguments
  - Returns a properly formatted tool message for the messages array
  - Returns a descriptive error message (not a Python exception) if anything
    fails — the LLM receives the error and can explain it to the user

This module is intentionally decoupled from agent.py so you can swap the
registry without changing the orchestration loop.

See docs/02-the-agent-loop/01-anatomy-of-an-agent.md — "The Hands (Tools)"
"""

from __future__ import annotations

import json
import logging
import time
from typing import Any, Callable

logger = logging.getLogger(__name__)

ToolRegistry = dict[str, Callable[..., Any]]


def dispatch_tool(
    tool_call: Any,
    registry: ToolRegistry | None = None,
) -> dict:
    """Execute a tool call and return a formatted tool message.

    Args:
        tool_call: An OpenAI ToolCall object with .id, .function.name,
                   and .function.arguments (JSON string).
        registry:  Mapping of tool names to callables. Defaults to the
                   built-in tools (get_weather, get_stock_price).

    Returns:
        A tool message dict ready to append to the messages array::

            {"role": "tool", "content": "<json string>", "tool_call_id": "<id>"}

        On error, the content is ``{"error": "<description>"}`` — the LLM
        receives this and can explain the failure to the user.
    """
    if registry is None:
        from tools import get_weather, get_stock_price
        registry = {
            "get_weather": get_weather,
            "get_stock_price": get_stock_price,
        }

    name: str = tool_call.function.name
    tool_call_id: str = tool_call.id

    # Parse arguments — the SDK delivers them as a JSON string.
    raw_args = tool_call.function.arguments
    try:
        args: dict = json.loads(raw_args) if isinstance(raw_args, str) else raw_args
    except json.JSONDecodeError as exc:
        logger.error("Tool %r: invalid argument JSON: %s", name, exc)
        return _error_message(tool_call_id, f"Invalid arguments JSON: {exc}")

    # Look up the tool.
    fn = registry.get(name)
    if fn is None:
        available = ", ".join(sorted(registry))
        logger.warning("Tool %r not found. Available: %s", name, available)
        return _error_message(
            tool_call_id,
            f"Tool '{name}' is not available. Available tools: {available}",
        )

    # Execute with timing.
    start = time.perf_counter()
    try:
        result = fn(**args)
    except TypeError as exc:
        elapsed = (time.perf_counter() - start) * 1000
        logger.error("Tool %r: argument mismatch (%.1f ms): %s", name, elapsed, exc)
        return _error_message(tool_call_id, f"Wrong arguments for '{name}': {exc}")
    except Exception as exc:  # noqa: BLE001
        elapsed = (time.perf_counter() - start) * 1000
        logger.error("Tool %r: execution failed (%.1f ms): %s", name, elapsed, exc)
        return _error_message(tool_call_id, f"Tool '{name}' failed: {exc}")

    elapsed_ms = (time.perf_counter() - start) * 1000
    result_json = json.dumps(result)
    preview = result_json[:200] + "…" if len(result_json) > 200 else result_json
    logger.debug(
        "Tool %r(%s) → %s  [%.1f ms]",
        name,
        json.dumps(args),
        preview,
        elapsed_ms,
    )

    return {
        "role": "tool",
        "content": result_json,
        "tool_call_id": tool_call_id,
    }


def _error_message(tool_call_id: str, error: str) -> dict:
    """Return a tool message that tells the LLM what went wrong."""
    return {
        "role": "tool",
        "content": json.dumps({"error": error}),
        "tool_call_id": tool_call_id,
    }
